pkg/graveler/committed/merge.go

```diff
@@ -0,0 +1,425 @@
+				if err := m.writeRange(iterRange); err != nil {
+					return err
+				}
+			}
+			if !iter.NextRange() {
+				break
+			}
+		} else {
+			baseValue, err := m.baseKeyGE(iterValue.Key)
+			if err != nil {
+				return fmt.Errorf("base value GE: %w", err)
+			}
+			if baseValue == nil || !bytes.Equal(baseValue.Identity, iterValue.Identity) {
+				if err := m.writeRecord(iterValue); err != nil {
+					return err
+				}
+			}
+			if !iter.Next() {
+				break
+			}
+		}
+	}
+	return iter.Err()
+}
+
+// applyBothRanges applies merging when both source and dest on range header
+func (m *merger) applyBothRanges(sourceRange *Range, destRange *Range) error {
+	switch {
+	case sourceRange.ID == destRange.ID: // source and dest added the same range
+		err := m.writeRange(sourceRange)
+		if err != nil {
+			return err
+		}
+		m.haveSource = m.source.NextRange()
+		m.haveDest = m.dest.NextRange()
+
+	case bytes.Compare(sourceRange.MaxKey, destRange.MinKey) < 0: // source before dest
+		baseRange, err := m.baseRangeGE(graveler.Key(sourceRange.MinKey))
+		if err != nil {
+			return fmt.Errorf("base range GE: %w", err)
+		}
+		if baseRange != nil && sourceRange.ID == baseRange.ID { // dest deleted this range
+			m.haveSource = m.source.NextRange()
+			return nil
+		}
+		if baseRange == nil || destRange.ID == baseRange.ID { // source added this range
+			err = m.writeRange(sourceRange)
+			if err != nil {
+				return err
+			}
+			m.haveSource = m.source.NextRange()
+			return nil
+		}
+		return graveler.ErrConflictFound // conflict - both changed this range
+
+	case bytes.Compare(destRange.MaxKey, sourceRange.MinKey) < 0: // dest before source
+		baseRange, err := m.baseRangeGE(graveler.Key(destRange.MinKey))
+		if err != nil {
+			return fmt.Errorf("base range GE: %w", err)
+		}
+		if baseRange != nil && sourceRange.ID == baseRange.ID { // dest added this range
+			err = m.writeRange(destRange)
+			if err != nil {
+				return err
+			}
+			m.haveDest = m.dest.NextRange()
+			return nil
+		}
+		if baseRange == nil || destRange.ID == baseRange.ID { // source deleted this range
+			m.haveDest = m.dest.NextRange()
+			return nil
+		}
+		return graveler.ErrConflictFound // conflict - both changed this range
+
+	case bytes.Equal(sourceRange.MinKey, destRange.MinKey) && bytes.Equal(sourceRange.MaxKey, destRange.MaxKey): // same bounds
+		baseRange, err := m.baseRangeGE(graveler.Key(sourceRange.MinKey))
+		if err != nil {
+			return err
+		}
+		if baseRange != nil && (sourceRange.ID == baseRange.ID || destRange.ID == baseRange.ID) {
+			if sourceRange.ID == baseRange.ID { // dest added changes
+				err = m.writeRange(destRange)
+			} else {
+				err = m.writeRange(sourceRange) // source added changes
+			}
+			if err != nil {
+				return err
+			}
+			m.haveSource = m.source.NextRange()
+			m.haveDest = m.dest.NextRange()
+		} else { // enter both ranges
+			m.haveSource = m.source.Next()
+			m.haveDest = m.dest.Next()
+		}
```

Comment:
IMO, more readable like this, but it's matter of preference:
```suggestion
		if baseRange == nil {
			fallthrough;
		}
		if sourceRange.ID == baseRange.ID {
			err = m.writeRange(destRange)
		} else if destRange.ID == baseRange.ID {
			err = m.writeRange(sourceRange)
		} else {
			fallthrough;
		}
		if err != nil {
			return err
		}
		m.haveSource = m.source.NextRange()
		m.haveDest = m.dest.NextRange()
```